custom/middleware/api.js

```diff
@@ -0,0 +1,96 @@
+'use strict'
+
+const router = require('express-promise-router')()
+const {Datastore} = require('@google-cloud/datastore')
+const {getAuth} = require('../../server/auth')
+const log = require('../../server/logger')
+
+const apiRoutes = {
+  '/api/upvote': handleVote(1),
+  '/api/downvote': handleVote(-1)
+}
+
+async function handleApi(req, res, next) {
+  const handler = apiRoutes[req.path]
+
+  if (typeof handler === 'function') {
+    handler(req, res, next)
+  } else {
+    next()
+  }
+}
+
+function handleVote(vote) {
+  return async function (req, res, _) {
+    const documentId = req.body.id
+    const reason = req.body.reason
+    const userInfo = req.userInfo
+    const datastoreClient = await getDatastoreClient()
+
+    recordVote(documentId, userInfo, vote, reason, datastoreClient)
+
+    res.send({documentId})
+  }
+}
+
+router.use(handleApi)
+
+/* Support Functions */
+
+async function getDatastoreClient() {
+  const projectId = process.env.GCP_PROJECT_ID
+  if (!projectId) {
+    log.warn('No GCP_PROJECT_ID provided! Will not connect to GCloud Datastore!')
+    return null
+  }
+
+  // because auth credentials may be passed in multiple ways, recycle pathway used by main auth logic
+  const {email, key} = await getAuth()
+
+  return new Datastore({
+    projectId,
+    credentials: {
+      client_email: email,
+      private_key: key
+    }
+  })
+}
+
+async function recordVote(documentId, userInfo, vote, voteReason, datastoreClient) {
+  const docKey = datastoreClient.key(['LibraryViewDoc', [userInfo.userId, documentId].join(':')])
+  updateVoteRecord(docKey, {documentId, vote, voteReason}, userInfo, datastoreClient)
+}
+
+function updateVoteRecord(viewKey, metadata, userInfo, datastoreClient) {
+  datastoreClient.get(viewKey)
+    .then((results) => {
+      const existing = results[0]
+      let updatedData
+
+      if (existing) {
+        updatedData = existing
+        updatedData.vote = metadata.vote
+        updatedData.voteDate = new Date()
+        updatedData.voteReason = metadata.voteReason
+      } else {
+        updatedData = Object.assign({
+          userId: userInfo.userId,
+          email: userInfo.email,
+          voteDate: new Date()
+        }, metadata)
```

Comment:
re: adding entity voting. did that require any update to the datastore model or its indexes? or does storing that extra data work implicitly through this code?

Reply:
We didn't add a new entity but just added new properties to the existing `LibraryViewDoc` entity.  Click [here](https://console.cloud.google.com/datastore/entities;kind=LibraryViewDoc;ns=__$DEFAULT$__/query/kind?project=content-library-development&supportedpurview=project) to check it out.

If you take a look at the README on this PR, I talk a little bit about it too.  No new indexes were added, mainly just added new properties to an existing entity.

Reply:
I didn't add any indexes to `LibraryViewDoc`.  All that was done was new properties were added to the entity.  I saw indexes can be created but I didn't think it was necessary for this property.  The README should explain more and the reasoning behind the chosen values, etc.